### PR TITLE
[Xamarin.Android.Build.Tasks] ILRepack SRM

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Downstream in monodroid, I attempted to use the final pkg file for the
Xamarin.Android commercial product.

Unfortunately, building a Xamarin.Android project fail with:

    $ msbuild HelloForms/HelloForms.Android/HelloForms.Android.csproj /t:SignAndroidPackage /restore /bl
    ...
    "HelloWorld/HelloForms/HelloForms.Android/HelloForms.Android.csproj" (SignAndroidPackage target) (1:7) ->
    (_ResolveAssemblies target) ->
    Xamarin.Android.Common.targets(1930,2): error : Exception while loading assemblies: System.MissingMethodException: Method not found: System.Reflection.AssemblyName System.Reflection.Metadata.AssemblyDefinition.GetAssemblyName() [HelloWorld/HelloForms/HelloForms.Android/HelloForms.Android.csproj]
    Xamarin.Android.Common.targets(1930,2): error :   at Xamarin.Android.Tasks.ResolveAssemblies.AddAssemblyReferences (Xamarin.Android.Tasks.MetadataResolver resolver, System.Collections.Generic.Dictionary`2[TKey,TValue] assemblies, System.String assemblyPath, System.Collections.Generic.List`1[T] resolutionPath) [0x0003b] in <41aa6ba2d5de4e9284e4372b8c641c57>:0  [HelloWorld/HelloForms/HelloForms.Android/HelloForms.Android.csproj]
    Xamarin.Android.Common.targets(1930,2): error :   at Xamarin.Android.Tasks.ResolveAssemblies.Execute (Xamarin.Android.Tasks.MetadataResolver resolver) [0x001bf] in <41aa6ba2d5de4e9284e4372b8c641c57>:0  [HelloWorld/HelloForms/HelloForms.Android/HelloForms.Android.csproj]

        1 Error(s)

It seems that MSBuild has its own `System.Reflection.Metadata.dll`:

    /Library/Frameworks/Mono.framework/Versions/Current/lib//mono/msbuild/15.0/bin/System.Reflection.Metadata.dll

It appears that this assembly is getting loaded instead of ours.

To fix this, I added the two assemblies needed by
`System.Reflection.Metadata` to `ILRepack.targets`.

After building a new "IL merged" `Xamarin.Android.Build.Tasks.dll` and
putting it into my system Xamarin.Android install, it fixed the
problem.

## Other changes ##

In c22475d, I removed a reference to `System.Reflection.Metadata` in
`xabuild.csproj` (part of MSBuild).

We need this again, since it breaks the build on macOS to remove it. I
suspect it was working in the interim, because we had our own copy of
`System.Reflection.Metadata` that is now IL merged.

### CODE REVIEW

**Comment by jonpryor** [comment]
The [**macOS PR Release Build**](
https://jenkins.mono-project.com/job/xamarin-android-pr-builder-release/460/) failed:

```
.../xamarin-android/bin/Release/lib/xamarin.android/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(2193,5): error MSB4018: System.IO.FileNotFoundException: Could not load assembly 'CodeBehindBuildTests, Version=0.0.0.0, Culture=neutral, PublicKeyToken='. Perhaps it doesn't exist in the Mono for Android profile? [/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/tests/CodeBehind/BuildTests/CodeBehindBuildTests.csproj]
```

This error doesn't make sense to me in the context of this PR, so I've restarted the build.

**Comment by jonathanpeppers** [comment]
Ok, latest build failure:
```
"/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/build-tools/create-vsix/create-vsix.csproj" (default target) (1) ->
(CreateVsixContainer target) -> 
  /Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/packages/Microsoft.VSSDK.BuildTools.15.0.26201/tools/VSSDK/Microsoft.VsSDK.targets(590,5): error VSSDK1025: Could not add the file "../../bin/Debug/lib/xamarin.android/ThirdPartyNotices.txt" to the zip package "bin/Debug/Xamarin.Android.Sdk-OSS-9.1.199.109_HEAD_1c4ac31.vsix". Could not find file "/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/bin/Debug/lib/xamarin.android/ThirdPartyNotices.txt" [/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/build-tools/create-vsix/create-vsix.csproj]

    2 Warning(s)
    1 Error(s)
```

How is `bin/Debug/lib/xamarin.android/ThirdPartyNotices.txt` missing...

**Comment by jonathanpeppers** [comment]
Wait...
```
+ make create-vsix CONFIGURATIONS=Release
...
/Library/Frameworks/Mono.framework/Versions/5.18.0/lib/mono/msbuild/15.0/bin/MSBuild.dll /binaryLogger:/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android//bin/BuildDebug/msbuild-20190118T151251-create-vsix.binlog /p:Configuration=Debug /p:Configuration=Debug /p:CreateVsixContainer=True /v:normal build-tools/create-vsix/create-vsix.csproj
```
Why is `CONFIGURATIONS=Release` building a `Debug` vsix? `bin/Debug/lib/xamarin.android/ThirdPartyNotices.txt` is missing because nothing `Debug` was ever built.

**Comment by jonathanpeppers** [comment]
I updated the PR builder job configuration to just do:
```
+ make create-vsix CONFIGURATION=Release
```
`$(CONFIGURATIONS)` isn't used anymore.

File: tools/xabuild/xabuild.csproj [hunk 1]
```diff
@@ -37,6 +37,9 @@
     <Reference Include="MSBuild">
       <HintPath>$(MSBuildReferencePath)\MSBuild.$(_MSBuildExtension)</HintPath>
     </Reference>
+    <Reference Include="System.Reflection.Metadata" Condition=" '$(OS)' != 'Windows_NT' ">
```
Comment (jonpryor) [inline tools/xabuild/xabuild.csproj]: What happens on Windows?  Why isn't this needed there?

> it breaks the build on macOS to remove [the `@(Reference)` to `System.Reflection.Metadata`].

What's the exact breakage?

`monodis --assemblyref bin/Debug/bin/xabuild.exe` doesn't show any references from `xabuild.exe` to `System.Reflection.Metadata.dll`, so I do not understand why this `@(Reference)` would be needed at all.

That said, I *do*  have a `bin/Debug/bin/System.Reflection.Metadata.dll`, so *something* copied it into `bin/Debug/bin` -- presumably this `@(Reference)` from earlier (I haven't cleaned `bin/Debug/bin` recently) -- but it's entirely possible that having `bin/Debug/bin/System.Reflection.Metadata.dll` is the actual problem...

Comment (jonathanpeppers) [reply tools/xabuild/xabuild.csproj]: It appears that `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\System.Reflection.Metadata.dll` does not exist on Windows, and none of the MSBuild assemblies reference it.

Comment (jonathanpeppers) [reply tools/xabuild/xabuild.csproj]: On macOS `MSBuild.dll` references `System.Reflection.Metadata.dll`.

Comment (jonathanpeppers) [reply tools/xabuild/xabuild.csproj]: After some discussion, trying the `AppDomain.CurrentDomain.AssemblyResolve` event.

**Comment by jonathanpeppers** [comment]
Error is:
```
/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/build-tools/scripts/TestApks.targets(128,5): warning MSB3073: The command "kill -KILL 64651" exited with code 1. [/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/tests/RunApkTests.targets]
...
"/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/build-tools/scripts/RunTests.targets" (RunAllTests target) (1) ->
(RunAllTests target) -> 
  /Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/build-tools/scripts/RunTests.targets(158,5): error : Execution of target RunApkTests exited with code 1.

    11261 Warning(s)
    1 Error(s)
```
Same happened on this PR: https://github.com/xamarin/xamarin-android/pull/2626

File: tools/xabuild/XABuild.cs [hunk 2]
```diff
@@ -15,6 +16,16 @@ static int Main ()
 		{
 			var paths = new XABuildPaths ();
 			try {
+				//HACK: on Mono + macOS, MSBuild cannot resolve System.Reflection.Metadata
+				AppDomain.CurrentDomain.AssemblyResolve += (sender, args) => {
```
Comment (jonpryor) [inline tools/xabuild/XABuild.cs]: Should this be done on all platforms, or just macOS & Linux?

**Pull request merged**